File: src/common/button.rb
require File.join(File.dirname(__FILE__), "..", "util", "actions")

class Button
  
  include Actions::ActiveRectangleSubscriber
  
  def initialize(owner, view_options, callback_name = nil, &callback)
    @x, @y, @z = view_options[:x], view_options[:y], view_options[:z]
    @dx, @dy = view_options[:dx], view_options[:dy]
    @owner = owner
    @window = @owner.is_a?(BakeryWizard::Window) ? @owner : @owner.window
    @callback = block_given? ? callback : callback_name
    @body = Gosu::Image.new(@window.window, "media/#{view_options[:image] || callback_name}_button.png", true)
  end
  
  def activate
    @window.register(self)
  end
  
  def deactivate
    @window.unregister(self)
  end
  
  def handle(event)
    sleep(0.2)
    @callback.is_a?(Symbol) && @owner.send(@callback)
    @callback.is_a?(Proc) && @callback.call
  end

  def render
    @body.draw(@x, @y, zindex)
  end

  def zindex
    @z
  end

  protected
  def active_x
    return @x, @x + @dx
  end

  def active_y
    return @y, @y + @dy
  end
end

File: src/common/button.py
import pygame
from util import game_util, actions
import zorder
import surface_sprite

class Button(surface_sprite.SurfaceSprite, actions.ActiveRectangleSubscriber):
    def __init__(self, owner, callback, publisher, x, y, image_path = 'default.png', layer = zorder.BUTTONS):
        surface_sprite.SurfaceSprite.__init__(self, x, y, image_path, layer)
        actions.ActiveRectangleSubscriber.__init__(self, x, y, self.rect.width, self.rect.height)
        self.__callback = getattr(owner, callback)
        self.__publisher = publisher
        
    def handle(self, event):
        self.__callback()

    def activate(self):
        self.__publisher.register(self)

    def deactivate(self):
        self.__publisher.unregister(self)

        
